EIPS/eip-6404.md

```diff
@@ -0,0 +1,1000 @@
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### SSZ
+
+#### `Optional[T]`
+
+A new [SSZ type](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md) is introduced to represent `Optional[T]` values.
+
+- If value is `None`, serialize as `[]`, and merkleize as `List[T, 1]` with length `0`.
+- If value is not `None`, serialize as `T`, and merkleize as `List[T, 1]` with length `1`.
+- Serialize `Optional[T]` as variable-size object (create offset-table entry in enclosing containers).
+
+#### Fixed capacity `Container` (TBD)
+
+The merkleization for SSZ `Container` structures generates a tree of minimum depth to contain all leaves. This is problematic whenever the number of leaves changes to require a different power of 2, as all `GeneralizedIndex` values change.
+
+SSZ `Container` structures should be annotated with a maximum capacity, similar to SSZ `List`. Unused space would be zero-extended accordingly.
+
+TBD
+
+### Transactions
+
+This design proposes a single normalized `Transaction` type. To include a transaction from an earlier type into a block, it needs to be migrated using `upgrade_signed_rlp_transaction_to_ssz`. For the purpose of signature validation, the originally signed transaction hash can be recomputed using `get_transaction_sighash`. For applications that need a perpetual identifier, `get_transaction_id` can be used.
+
+#### Hash version
+
+For the purpose of transaction signature validation, it is necessary to recover the originally signed transaction hash. A new enumeration is introduced to indicate the original transaction hashing method. Note that this is different from the [EIP-2718](./eip-2718.md) transaction type, as legacy transactions may have been signed using different methods. The [EIP-2718](./eip-2718.md) transaction type is still used to encapsule different transaction formats on the network.
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `TxHashVersion` | `uint8` | Original hashing method used for signing the transaction
+
+| Name | Value | Description |
+| - | - | - |
+| `TX_HASH_VERSION_LEGACY` | `TxHashVersion(0x00)` | Replayable [`LegacyTransaction`](./eip-2718.md#transactions) |
+| `TX_HASH_VERSION_EIP155` | `TxHashVersion(0x01)` | [EIP-155](./eip-155.md) `LegacyTransaction` |
+| `TX_HASH_VERSION_EIP2930` | `TxHashVersion(0x02)` | [EIP-2930](./eip-2930.md#definitions) transaction |
+| `TX_HASH_VERSION_EIP1559` | `TxHashVersion(0x03)` | [EIP-1559](./eip-1559.md#specification) transaction |
+| `TX_HASH_VERSION_EIP4844` | `TxHashVersion(0x04)` | [EIP-4844](./eip-4844.md#parameters) transaction |
```

Comment:
4844 has `TransactionType = 0x05`. Shouldn't the `TxHashVersion`s here mirror the `TransactionType`?

Reply:
Not having it mirror will result in:
- the requirement to specify both `TxHashVersion` and `TransactionType` in new EIP-2718 extensions (like EIP-4844),
- the potential for easy mistakes in implementations.

I understand the reason it can't easily be the same, because we want to separate legacy and EIP-155, but maybe there's a solution somewhere.

Reply:
`TxHashVersion` and EIP-2718 transaction types, are, conceptually, different.

`TxHashVersion` solely indicates the hashing method that was originally used to sign the transaction and to derive it's perpetual ID. There is no meaning beyond that, it is not even part of the `Transaction` but instead a property of the `SignedTransaction`.

Multiple EIP-2718 tx types may map to a single `TxHashVersion`:
- EIP-2718 tx type 5 maps to `TxHashVersion` 4
- EIP-2718 tx type 2 maps to `TxHashVersion` 3
- EIP-2718 tx type 1 maps to `TxHashVersion` 2
- EIP-2718 tx types 0xc0 through 0xef map to `TxHashVersion` 1 or 0, depending on the signature's `v` value
- EIP-2718 tx types 0, 3, 4, 0x80 through 0xbf, and 0xff are reserved

I agree that it would be cool to have a 1:1 mapping from EIP-2718 to `TxHashVersion`. Maybe there is a way to at least ensure that the 1/2/5 values match? But in the end, it is something that can be tested very easily.

Reply:
I know they are different conceptually, the question is if we can avoid making them different to avoid future confusion.

> it is something that can be tested very easily.

Everything in the system can be tested, yet we see numerous consensus bugs popping up in precompiles which have been live for years 🤓

Reply:
That's true as well 😅

The only function depending on ordering is `validate_transaction_payload`. If that property should be preserved, we could either change the `BLOB_TX_TYPE` in EIP-4844 to `0x04`, or, alternatively, `TxHashVersion` could start from `1` instead of `0`. This would then at least ensure that going forward, `TxHashVersion` matches EIP-2718 transaction type, essentially ensuring that there is no ongoing maintenance burden from this.

Alternatively, if `TxHashVersion` ordering is not desirable, we could use values 3 and 4 to indicate EIP-155 and legacy, so that 1 and 2 can remain EIP-2930 / EIP-1559.

Reply:
Just checked JSON-RPC. There, it seems that type `0x00` is indicated for both legacy and EIP-155 transactions.
Is there a spare, reserved, `chain_id` value that could be used to distinguish explicit replayability (EIP-155 `chain_id=0`) from implicit replayability (pre-EIP-155 tx)?

Technically, (uint64) `0x7ffffffffffffffc` ends up at 27/28 when applying `v = chain_id * 2 + 35 + {0,1}`.
But seems quite ugly, as well.

In JSON-RPC, the output is the `v` value, not `y_parity`. That could also help, but is kinda ugly as it leaks chain ID specifics into a general-purpose `ECDSASignature` type.


JSON-RPC examples:

```json
[
      {
        "accessList": [],
        "blockHash": "0xeeb1a6c504394b835b06eda9288446d4739d2f70290f52e904be9b550347a2a1",
        "blockNumber": "0x2b0213",
        "chainId": "0xaa36a7",
        "from": "0xd123b238958ea02934e82436c5248dae5501ffd9",
        "gas": "0x5208",
        "gasPrice": "0x59682f07",
        "hash": "0xad0a532e17955b56d42663fab13f90c8310b3003a9ea706387803eba679d2c54",
        "input": "0x",
        "maxFeePerGas": "0x59682f0e",
        "maxPriorityFeePerGas": "0x59682f00",
        "nonce": "0x144e",
        "r": "0xbaba866770cfdb06cd1e912a3947c11e87d5d699489533c14a946515002f0ebc",
        "s": "0x4d9a0246253634bab7dd4663fa2863251603b966d2a05e938b48a067a8071f58",
        "to": "0xd0266a7e4f388c6e6f0cf11d2459624870acc7c2",
        "transactionIndex": "0x4",
        "type": "0x2",
        "v": "0x1",
        "value": "0x16345785d8a0000"
      },
      {
        "blockHash": "0xeeb1a6c504394b835b06eda9288446d4739d2f70290f52e904be9b550347a2a1",
        "blockNumber": "0x2b0213",
        "chainId": "0xaa36a7",
        "from": "0xe2a3422f3168149ad2d11b4de2b97b05f1ebf76e",
        "gas": "0x13880",
        "gasPrice": "0x3b9aca00",
        "hash": "0x1d7e73fc2002a9b02cbab28bd295540f982a16ed10e2a75fc804ff1b59982183",
        "input": "0xa9059cbb000000000000000000000000e7d14b98386cb0752dfca69151f85c91eaa87b540000000000000000000000000000000000000000000000000de0b6b3a7640000",
        "nonce": "0x10409",
        "r": "0x804d47cf511fe2f4c2503e2aeeeab02353cad51566c568d2a3cb97d17cff619e",
        "s": "0x176e47941e0c57d7f0898138b165bf8cf3d29a2b9170b63dc207e0cfef235ea3",
        "to": "0xed43f81c17976372fcb5786dd214572e7dbb92c7",
        "transactionIndex": "0x5",
        "type": "0x0",
        "v": "0x1546d71",
        "value": "0x0"
      }
]
```

Reply:
I have updated the EIP to use matching 0x01 / 0x02 / 0x05 EIP-2718 tx types, by using 0x03 / 0x04 for the legacy types. I have put a section about reserved types, also mentioning the 0x00 value on JSON-RPC to prevent future accidents.